src/kemal/middleware/csrf.cr

```diff
@@ -0,0 +1,46 @@
+require "secure_random"
+require "http"
+
+module Kemal::Middleware
+  # This middleware adds CSRF protection to your application.
+  #
+  # Returns 403 "Forbidden" unless the current CSRF token is submitted
+  # with any non-GET/HEAD request.
+  #
+  # Without CSRF protection, your app is vulnerable to replay attacks
+  # where an attacker can re-submit a form.
+  #
+  class CSRF < HTTP::Handler
+    HEADER = "HTTP_X_CSRF_TOKEN"
+    ALLOWED_METHODS = %w[GET HEAD OPTIONS TRACE]
+    PARAMETER_NAME = "authenticity_token"
+
+    def call(context)
+      unless context.session["csrf"]?
+        context.session["csrf"] = SecureRandom.hex(16)
+      end
+
+      return call_next(context) if ALLOWED_METHODS.includes?(context.request.method)
+
+      req = context.request
+      submitted = if req.headers[HEADER]?
+        req.headers[HEADER]
+      elsif context.params.body[PARAMETER_NAME]?
+        context.params.body[PARAMETER_NAME]
+      else
+        "nothing"
+      end
+      current_token = context.session["csrf"]
+
+      if current_token == submitted
+        # reset the token so it can't be used again
+        context.session["csrf"] = SecureRandom.hex(16)
```

Comment:
I noticed a [rack implementation](https://github.com/baldowl/rack_csrf/blob/master/lib/rack/csrf.rb#L59) was using urlsafe_base64. I see you are limiting it to just body parameters (which is probably a good thing) so I guess it won't matter much.